src/components/game/graphics/objects/board/Board.js

```diff
@@ -0,0 +1,30 @@
+import { drawCell } from "../cell/DrawCell";
+import { updateCell } from "../cell/UpdateCell";
+
+export const renderBoard = (canvas, board, xCellSize, yCellSize) => {
+// Draw grid cells
+  var [xSize, ySize] = [board[0].length, board.length];
+    for (var y = 0; y < ySize; y++) {
+      for (var x = 0; x < xSize; x++) {
+        drawCell(
+          canvas,
+          board[y][x],
+          x * xCellSize,
+          y * yCellSize,
+          xCellSize,
+          yCellSize
+        );
+      }
+    }
+}
+
+export const updateBoard = (board) => {
+    // Update grid cells
+    var [xSize, ySize] = [board[0].length, board.length];
+    for (var y = 0; y < ySize; y++) {
+      for (var x = 0; x < xSize; x++) {
+        updateCell(board[y][x]);
```

Comment:
Unmark cells at updateCell instead of in GameState and see if that works, and make sure updateBoard is called after render; this will save a whole traversal through the board.

Reply:
better to unmark separately to keep path marking decoupled from the core state. leave as is